### ISSUE TITLE
[CVE-2023-27043] Parsing errors in email/_parseaddr.py lead to incorrect value in email address part of tuple

### ISSUE THREAD

**Issue by tdwyer** [python/cpython#102988]
# Bug report

I have discovered a way to cause `email.utils.parsaddr()` and `email.utils.getaddresses()` to erroneously return the Real Name portion of a RFC 2822 Address Header in the Email Address portion of the returned tuple. This vulnerability enables me to bypass security systems which allow or deny access or actions based on the email address.

Example one, this is an issue if the system should only allow for accounts to be created for email addresses which are part of a whitelist e.g. only government emails, or a blacklist e.g. block `attacker@example.com`. If ether `parseaddr()` or `getaddresses()` is used to parse the Address Header and the Email Address part of the returned tuple is checked ageist the whitelist/blacklist, it would be checking the value in the Real Name part of of the Address Header instead of the Email Address. Then, to finish account registration the system sends the reply to the full Address Header, which is normally the case, the reply would be sent to the actual Email Address and not the Real Name part of the header.

This is similar to the type of attack which was used to exploit the French governments Tchap chat system a few years back.

Tchap: The super (not) secure app of the French government
https://medium.com/@fs0c131y/tchap-the-super-not-secure-app-of-the-french-government-84b31517d144

Example two, this would also be an issue if an email security system is written in Python and uses `parseaddr()` or `getaddresses()` to parse the Address Header of incoming email and then checks the Email Address part of the tuple against a whitelist/blacklist. I can use this vulnerability to trick that system into checking the Real Name part of the Address Header instead of the actual Email Address which would have been blocked. Then after this, the email system which actually replies to the email dose not use Python's `parseaddr()` or `getaddresses()` functions to get the email address, or simply uses the full Address Header to send a reply.

For all of these test cases Golang returns a Parsing Error, but all versions of Python including Python 2, Python 3, and the most resent version in Github all result in the Real Name part of the address being returned in the Email Address part of the tuple.

**NOTE: This issue has been assigned: CVE-2023-27043
Link: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2023-27043**

**[VulnerabilityType Other]
CWE-1286: Improper Validation of Syntactic Correctness of Input**

I have been in communication with security@python.org about this issue, and was told they are fine with addressing this issue publicly. Additionally, I have created a PR which will detect these parsing errors and return an empty tuple to indicate a parsing error instead of returning the erroneous output.

Example Script to show this erroneous output:
```
from email.utils import parseaddr
rfc2822_ADDRESS = "Thomas Dwyer<tdwyer@example.org>"
print("Valid RFC2822 Address Parse Output")
parseaddr(rfc2822_ADDRESS)
print()
print("Test output...")
print()
specials = '()<>@,:;.\"[]'
a = "alice@example.org"
b = "<bob@example.org>"
for i in specials:
    c = a + i + b
    print(c)
    parseaddr(c)
```

Example Golang program which shows that Golang returns a parsing error:
```
package main

import (
 "fmt"
 "net/mail"
)

func main() {
   fmt.Println((&mail.AddressParser{}).Parse("Thomas Dwyer <thomas@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("thomas@example.org"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org(<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org)<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org<<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org><bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org@<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org,bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org:<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org;<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org.<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse(`alice@example.org"<bob@example.org>`))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org[<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org]<bob@example.org>"))
   fmt.Println((&mail.AddressParser{}).Parse("alice@example.org<bob@example.org>"))
}
```

# Your environment

- CPython versions tested on: All versions of Python

### Linked PRs
* gh-102990
* gh-105127
* gh-105128
* gh-106612
* gh-108250
* gh-111116
* gh-123766
* gh-123767
* gh-123768
* gh-123769
* gh-123770
* gh-123771
* gh-123772

**Comment by bitdancer** [comment]
These addresses are problematic.  Take, for example, alice@example.org(<bob@example.org>.  The new email parser views this as a valid address (alice@example.org) followed by an incomplete comment, which is the *correct* parsing of that string per the RFCs.  alice@example.org@<bob@example.org> on the other hand is treated as an invalid address because of the extra @.  I haven't actually looked at the others to see what the new parser does, but in general these look like *valid* addresses followed by junk (some of which also happens to be a valid address).

In general, as has been pointed out repeatedly, if you depend on the content of the address fields of an email message for security you *are* going to get burnt; we can't fix that.

I personally don't care a lot about the old API, but "fixing" this could break applications, since addresses that were previously considered valid (and arguably are valid in at least one case) will be rejected.  I think the risk is low, but the benefit is low as well.  In the new API, these should all have non-empty defects (might want to check that), so running with policy(raise_on_defect=True) or checking the defects attribute of the header will give you as much protection as there is to be had.

Now, you will note that you can easily put "From: Alice <alice@example.org>" as the from header on a message that did not come from alice@examp.org.  So this fix does not provide any useful protection against your example 2.  Being able to exploit "put the real address in the display name, some junk, and my address at the end" requires that the email system doing the replying treat the address differently than python and chooses to use the trailing part of the invalid address as the one to send to rather than the leading *valid* address that is followed by junk.  How likely is that?  Have you run tests?  Do you have examples of this occurring?  I suppose an email system could choose to look first at the address in angle brackets, especially if it is regex based rather than a standard RFC tokenizing parser.

So, I can see an argument to be made that we shouldn't apply postel's rule (be liberal in what you accept) so strongly here, and should in fact treat any address that does not end in a valid character as invalidating that address.  I don't know what the backward compatibility consequences of that would be, but if you implement this fix maybe we'll find out.  If it turns out not to be a problem, we could change the rules in the new parser as well, and invalidate the address rather than just registering a defect (for most of these examples, at least).

**Comment by tdwyer** [comment]
Hello, I agree that not all of the cases being checked in the example python script are a problem. In the case of adding a `,`, that is actually how a list of addresses is delimitated so for `getaddresses()` to return 2 address tuples is valid. However, `parseaddr()` is only for parsing a single address. So much so, that it is hard coded to only return the first element of the list returned by the parser `return addrs[0]`. The main solution for this in `parseaddr()` in my PR is to check if the returned list of parsed addresses is longer than one. The proposed solution in my PR for `getaddresses()` is to count the number of addresses which are in the input to be the same as the number of addresses returned by the parser. Additionally, it will check for a `[` in the returned email address and replace that with an empty tuple instead.

For all of the cases being checked, they all cause the parser to return both alice address and the bob address in the email address part of their own tuple. So, if the `alice@example.com(bob@example.org` is a valid address followed by an incomplete comment, why should it return the incomplete comment in the email address part of it's own tuple?

`[('', 'alice@example.com'), ('', 'bob@example.com')]`

My feeling is that if what is put in the email address part of the tuple is not a valid email address, it should return a parsing error instead.

Golang will return a parsing error indicating that when these headers are parsed they are returning more than a single address.

```"Thomas Dwyer" <thomas@example.org> <nil>
<thomas@example.org> <nil>
<nil> mail: expected single address, got "<notme@gmail.com>"
<nil> mail: misformatted parenthetical comment
<nil> mail: expected single address, got ")<notme@gmail.com>"
<nil> mail: expected single address, got "<<notme@gmail.com>"
<nil> mail: expected single address, got "><notme@gmail.com>"
<nil> mail: expected single address, got "@<notme@gmail.com>"
<nil> mail: expected single address, got ",notme@gmail.com>"
<nil> mail: expected single address, got ":<notme@gmail.com>"
<nil> mail: expected single address, got ";<notme@gmail.com>"
"tdwyer@example.org." <notme@gmail.com> <nil>
<nil> mail: expected single address, got "\"<notme@gmail.com>"
<nil> mail: expected single address, got "[<notme@gmail.com>"
<nil> mail: expected single address, got "]<notme@gmail.com>"
<nil> mail: expected single address, got "<notme@gmail.com>"```

**Comment by bitdancer** [comment]
Look up Postel's rule.  The grammar doesn't recognize an incomplete comment as a *valid* comment, but that doesn't change the fact that syntactically it is an incomplete comment.  But I've already agreed that I think we probably shouldn't apply Postel's rule so strongly in this case.

Also, we never throw parsing errors, we only register defects.  This is a hard rule for the email parser, and there are many practical reasons for this.  What we'd be changing here is whether or not we extract valid addresses out of questionable strings or treat them as invalid addresses and not return them.  Which is what you are proposing, and I am in agreement, even though I worry a little bit about backward compatibility (breaking applications that currently handle the broken output of some mail programs that is not intentionally malicious).

**Comment by tdwyer** [comment]
First, I'd like to say, in my PR added even more checks to `Lib/test/test_email/test_email.py` which in check all of the wacky examples in RFC 2822 and all of those, as well as, all of the current tests pass. All correctly formatted addresses are still parsed correctly after my fix is applied.

PR: https://github.com/python/cpython/pull/102990

I'd also like to address the concerns about returning an empty 2-tuple `('', '')` to indicate a parsing error. The comment for the `def parseaddr(adds):` function in `Lib/email/utils.py` says that is the correct thing to do

Link: https://github.com/python/cpython/blob/57f8f9a66d4d0b5e590f7746a58136b3b45b1336/Lib/email/utils.py#L208

```
def parseaddr(addr):
    """
    Parse addr into its constituent realname and email address parts.

    Return a tuple of realname and email address, unless the parse fails, in
    which case return a 2-tuple of ('', '').
    """
```

This issue is actually caused by a couple of bugs.

1. The code should detect incorrect syntax and return a 2-tuple of `('', '')` to indicate that parsing has failed
2. There are 2 functions to fix, `parseaddr()` and `getaddresses()`. Both of these functions need to check for syntax errors. However, the `parseaddr()` also needs to check for invalid input. That function should only be passed in a single address to check. However, if it is given a list of addresses it is hard coded to just returns the first 2-tuple from the list. What it should do is return `('', '')` here too or maybe this one could actually return a real Error. In my PR, I have it check for this and return `('', '')`. This is part of what causes the issue in `parseaddr()` because when these invalid addresses are passed in, the bug causes these to be parsed as a list of addresses each with one part in the Email Address part of the tuple. The first 2-tuple of that list is the one which contains the Real Name part in the Email Address part of the 2-tuple.

The fix for CVE-2019-16056 which was very similar to this CVE-2023-27043 ... was to return an empty 2-tuple `('', '')`
Link: https://bugs.python.org/issue34155#msg347157

Code example
Link: https://github.com/python/cpython/blob/57f8f9a66d4d0b5e590f7746a58136b3b45b1336/Lib/email/_parseaddr.py#L410
```
elif self.field[self.pos] == '@':
    # bpo-34155: Don't parse domains with two `@` like
    # `a@malicious.org@important.com`.
    return EMPTYSTRING
```

Additionally, the code return `('', '')` if a fail domain is not returned
Link: https://github.com/python/cpython/blob/57f8f9a66d4d0b5e590f7746a58136b3b45b1336/Lib/email/_parseaddr.py#L391
```
if not domain:
    # Invalid domain, return an empty address instead of returning a
    # local part to denote failed parsing.
    return EMPTYSTRING
```

**Comment by vstinner** [comment]
I created https://python-security.readthedocs.io/vuln/email-parseaddr-realname.html to track fixes of this issue.

**Comment by tdwyer** [comment]
@gpshead I had to create a new PR because I messed up Git when trying to push the changes: https://github.com/python/cpython/pull/105127

**Comment by gpshead** [comment]
I merged the PR for main and 3.12beta4, I want to let that sit for a while and see if anything crops up during beta4 and rc1 before landing backports for 3.11 or earlier.

**Comment by gpshead** [comment]
https://github.com/python/cpython/issues/106669 appears to be fallout from this.

**Comment by mcepl** [comment]
I have so far backported this patch for SUSE to 3.10, 3.9, 3.8, 3.6, and even 3.4 (yes, we have it still in one old SLE), and it seems to me to be rather unremarkable. Aside from adjusting positions of hunks when applying patches, I haven’t need to do anything for test suite to pass.

Now, I will try 2.7 and we’ll see whether I would succeed.

**Comment by gpshead** [comment]
I recommend people do NOT backport this change. It breaks existing valid email address parsing, see the above linked issue.

**Comment by mcepl** [comment]
> I recommend people do NOT backport this change. It breaks existing valid email address parsing, see the above linked issue.

I know, and I am not releasing the patched packages yet, waiting on #106669 to be resolved.

**Comment by alittlesir** [comment]
@gpshead  hello, my brother, is there a fix plan for this vulnerability?

**Comment by gpshead** [comment]
nope.  someone needs to come up with a fix that doesn't break anything.

**Comment by hhucke** [comment]
On Tue, 8 Aug 2023, Gregory P. Smith wrote:

> nope.  someone needs to come up with a fix that doesn't break anything.

Hi folks.

Erm... You are all aware that this bug wears a CVE number? So it shouldn't 
take too long to fix this bug. Every day without - an appropriate - fix 
makes it more likely that someone finds a way to expoit this 
vulnerability.

And honestly: Read the appropriate RFCs - I'm willing to carve them out 
for you. The later ones are infact more detailed descriptions about what 
is "allowed" and what not than the younger ones - and parse accordingly.
You will probably catch more than before and what you don't catch is most 
probably nothing what you /could/ use as a working e-mail address.

And yes you might break /some/ programs/services. But shit happens. We 
have a german saying: Better an end with horror than horror without end.

Best regards,
     Henning

**Comment by mcepl** [comment]
> Hi folks. Erm... You are all aware that this bug wears a CVE number? So it shouldn't take too long to fix this bug. Every day without - an appropriate - fix makes it more likely that someone finds a way to expoit this vulnerability.

Exactly, so where is the patch you have prepared? We are eagerly waiting on it!

Do you know that Python is a project run and done by volunteers mostly in their spare time? I don’t think there is anybody paid exactly to work on the upstream code (or too few people for the size of the codebase), so it is never “you, user” and “them, developers”, but always “Don’t ask what the Python community can do for you, but ask what you can do for the Python community”.

**Comment by hhucke** [comment]
On Wed, 9 Aug 2023, Matěj Cepl wrote:

Hi Matěj,

>> Hi folks. Erm... You are all aware that this bug wears a CVE number? So it shouldn't take too long to fix this bug. Every day without - an appropriate - fix makes it more likely that someone finds a way to expoit this vulnerability.
>
> Exactly, so where is the patch you have prepared? We are eagerly waiting on it!
>
> Do you know that Python is a project run and done by volunteers mostly in their spare time? I don’t think there is anybody paid exactly to work on the upstream code (or too few people for the size of the codebase), so it is never “you, user” and “them, developers”, but always “Don’t ask what the Python community can do for you, but ask what you can do for the Python community”.

I'm with you despite the fact that I'm a systems and network admin since 
28 years and likely wouldn't write code as well as developers from 
nowadays without question are able to do and despite the fact that I also 
wrote and somethimes write code.

Nonetheless there is obviously already a patch - read the former messages 
in this thread - which isn't applied because doubts that it'll not break 
things on which I wrote my comment...

Regards
    Henning

**Comment by theta682** [comment]
> Nonetheless there is obviously already a patch - read the former messages in this thread - which isn't applied because doubts that it'll not break things on which I wrote my comment...

The patch was tried and there were regressions without doubts. That's why it was rolled back.

**Comment by frenzymadness** [comment]
I've been staring at this for a couple of hours now trying to improve the original patch to pass also for the new cases and it seems that my confusion comes from the regression test mentioned here https://github.com/python/cpython/issues/106669#issuecomment-1632107729 . The test expects `'"Bud, Person" <bperson@dom.ain>'` to be parsed into `('Bud Person', 'bperson@dom.ain')` but in fact the test in [Django](https://github.com/django/django/blob/d25f3892114466d689fd6936f79f3bd9a9acc30e/tests/mail/tests.py#L462) and [b4](https://git.kernel.org/pub/scm/utils/b4/b4.git/tree/tests/test___init__.py#n166) both expect that the comma stays there so the correct result should be `('Bud, Person', 'bperson@dom.ain')`.

**Comment by frenzymadness** [comment]
I have an idea of how to improve the original patch.

The problem with it was in the comma counting/checking logic:

```python
    n = 0
    for v in fieldvalues:
        n += v.count(',') + 1

    if len(result) != n:
        return [('', '')]
```

What if we improve it to ignore parts of the strings enclosed in `""` or `''`:

```python
    pattern = r'"[^"]*"|\'[^\']*\''
    n = 0
    for v in fieldvalues:
        v = re.sub(pattern, "", v)
        n += v.count(",") + 1

    if len(result) != n:
        return [('', '')]
```

I've tried it and tests in `Lib/test/test_email/test_email.py` are passing. I've also tried to rebuild Django and b4 with the patched Python and both succeeded. Is there any other important project you know about I can try to test with this change?

@tdwyer what do you think about it?

**Comment by tdwyer-wish** [comment]
Oh boy, I should have checked this page first. I just figured out the same solution.

Ya, just stripping it out with a regex and then counting the `,`'s works. I do need to count the `,`'s because `getaddressess()` pares RFC 2822 Headers and each header can have a list of addresses. The `,` is the deliminator for multiple addresses in a list. However, the `,` can also be included in a Real Name part as long as it is in Quotes.

I'm making a new PR at this moment.

**Comment by tdwyer** [comment]
🤦 I should have read this Issue again. I just figured out the same solution.

I do need to count the `,`'s because even though there is a List being passed into the `getaddresses()` function it's a list of RFC 2822 Headers and each Header can be a list of email addresses.

Commas `,`'s are what deliminate multiple addresses in the Header. However, a `,` can also be included in the Real Name part as long as it is in Quotes.

One thing I'm unsure about is if they have to be Double-Quotes or if Single-Quotes can also be uses. I'm not sure if I'm reading the RFC correctly but it looks like they will always be Double-Quotes?

NOTE: I am making a new PR at this moment.

**Comment by tdwyer** [comment]
New PR.

Note, I just have the regex looking for Double Quotes. Let me know if Single Quotes can be used also, and I'll make the change.

`gh-102988: Detect email address parsing errors and return empty tuple to indicate the parsing error (old API) #108250`
Link: https://github.com/python/cpython/pull/108250

**Comment by vstinner** [comment]
@bitdancer:
> These addresses are problematic. Take, for example, [alice@example.org](mailto:alice@example.org)([bob@example.org](mailto:bob@example.org). The new email parser views this as a valid address ([alice@example.org](mailto:alice@example.org)) followed by an incomplete comment, which is the correct parsing of that string per the RFCs. alice@example.org@[bob@example.org](mailto:bob@example.org) on the other hand is treated as an invalid address because of the extra @. I haven't actually looked at the others to see what the new parser does, but in general these look like valid addresses followed by junk (some of which also happens to be a valid address).

The PR #108250 changes the behavior without providing a way to get the old behavior.

You may think that the old behavior is bad/invalid, but we had here for many years, applications are relying on production on this behavior.

So I wrote PR #111116 which is a copy of PR #108250 but **adds a strict=True parameter**. So it still rejects invalid arguments by default, but it gives an option to get the old "invalid" behavior: strict=False.

**Comment by zhuofeng6** [comment]
hello, my brother, the master branch has been fixed. 
when will the other branches of this cve be fixed?

**Comment by vstinner** [comment]
>  when will the other branches of this cve be fixed?

The plan is to wait for a few Python 3.13 releases before deciding to backport the change to stable branches. See the rationale for not fixing the issue in other branches yet in this discussion: https://github.com/python/cpython/pull/111116#issuecomment-1856451531

**Comment by ambv** [comment]
I don't know exactly how *long* we intend to wait, so how about this:
- in May there's Python 3.13 beta 1, which will get much more attention than the alphas are;
- we do backports with Python 3.13 beta 2 as long as no issues are identified?

**Comment by chandu131097** [comment]
Hi team, could you please tell me if this bug is fixed in python 3.12.2 release ? If not, when shall we expect the fix and in which release ?

**Comment by frenzymadness** [comment]
> Hi team, could you please tell me if this bug is fixed in python 3.12.2 release ?

It's not. The fix is in the main branch only (3.13). I don't have an answer for the second question.

**Comment by vstinner** [comment]
> we do backports with Python 3.13 beta 2 as long as no issues are identified?

Python 3.13 beta 2 was released. Is it ok to backport this change to other branches yet?

**Comment by markus-gerstel-sp** [comment]
Backport for 3.12 (https://github.com/python/cpython/pull/1066129) has landed, fix will be in 3.12.5

**Comment by seanthegeek** [comment]
I'm very late to this, but rather than refusing to parse the address `alice@example.com <bob@example.com>`, because emails with that address format could still be delivered by others, I would prefer the output `[('alice@example.com', 'noreply@example.com')]`
